File: guardian_sdk/detectors.py
from __future__ import annotations

import ipaddress
import re
from dataclasses import dataclass
from typing import Iterable, Optional, Tuple


@dataclass(frozen=True)
class Detection:
    category: str
    verdict: str  # "SUSPICIOUS" | "BLOCKED" | "INFO"
    severity: int  # 1-10
    reason: str
    indicators: Tuple[str, ...] = ()


SQLI_REGEXES: Tuple[Tuple[str, re.Pattern[str]], ...] = (
    ("union-select", re.compile(r"\bunion\b\s+\bselect\b", re.IGNORECASE)),
    ("tautology", re.compile(r"(\bor\b|\band\b)\s+[\w'\"]+\s*=\s*[\w'\"]+", re.IGNORECASE)),
    ("comment-seq", re.compile(r"(--|#|/\*)", re.IGNORECASE)),
    ("stacked", re.compile(r";\s*(drop|alter|create|insert|update|delete)\b", re.IGNORECASE)),
    ("sleep", re.compile(r"\b(sleep|pg_sleep|benchmark)\s*\(", re.IGNORECASE)),
)


def detect_sqli(query: str) -> Optional[Detection]:
    q = (query or "").strip()
    if not q:
        return None

    hits: list[str] = []
    for name, rx in SQLI_REGEXES:
        if rx.search(q):
            hits.append(name)

    if not hits:
        return None

    # Include "sql injection" literal so today's SOC heuristic scoring picks it up.
    return Detection(
        category="sql_injection",
        verdict="SUSPICIOUS",
        severity=7,
        reason="sql injection pattern detected",
        indicators=tuple(hits) + ("sql injection",),
    )


PRIVATE_HOST_LITERALS = ("localhost", "127.0.0.1", "0.0.0.0")
METADATA_IPS = ("169.254.169.254",)  # cloud instance metadata (AWS/Azure/GCP variants)


def _is_ip_private(ip_s: str) -> bool:
    try:
        ip = ipaddress.ip_address(ip_s)
        return bool(ip.is_private or ip.is_loopback or ip.is_link_local)
    except ValueError:
        return False


def detect_ssrf(url: str) -> Optional[Detection]:
    u = (url or "").strip()
    if not u:
        return None

    u_lc = u.lower()

    # Very cheap checks first.
    if u_lc.startswith(("file://", "gopher://", "ftp://")):
        return Detection(
            category="ssrf",
            verdict="SUSPICIOUS",
            severity=8,
            reason="dangerous outbound scheme (possible SSRF)",
            indicators=("ssrf", "outbound request"),
        )

    if any(h in u_lc for h in PRIVATE_HOST_LITERALS):
        return Detection(
            category="ssrf",
            verdict="SUSPICIOUS",
            severity=8,
            reason="outbound request to localhost (possible SSRF)",
            indicators=("ssrf", "localhost"),
        )

    for ip_s in METADATA_IPS:
        if ip_s in u_lc:
            return Detection(
                category="ssrf",
                verdict="SUSPICIOUS",
                severity=9,
                reason="outbound request to instance metadata (possible SSRF)",
                indicators=("ssrf", "metadata", ip_s),
            )

    # If URL contains an IPv4 literal, check whether it's private/link-local.
    m = re.search(r"\b(\d{1,3}(?:\.\d{1,3}){3})\b", u)
    if m and _is_ip_private(m.group(1)):
        return Detection(
            category="ssrf",
            verdict="SUSPICIOUS",
            severity=8,
            reason="outbound request to private IP (possible SSRF)",
            indicators=("ssrf", "private ip", m.group(1)),
        )

    return None


def summarize_indicators(d: Optional[Detection]) -> str:
    if not d:
        return ""
    if not d.indicators:
        return d.reason
    return f"{d.reason} indicators={list(d.indicators)}"


def clamp_severity(sev: int) -> int:
    try:
        v = int(sev)
    except Exception:
        v = 1
    return max(1, min(10, v))

File: guardian-sdk-node/src/detectors.js
class Detection {
    constructor(category, verdict, severity, reason, indicators = []) {
        this.category = category;
        this.verdict = verdict;
        this.severity = severity;
        this.reason = reason;
        this.indicators = indicators;
    }
}

const SQLI_REGEXES = [
    ["union-select", /\bunion\b\s+\bselect\b/i],
    ["tautology", /(\bor\b|\band\b)\s+[\w'"]+\s*=\s*[\w'"]+/i],
    ["comment-seq", /(--|#|\/\*)/i],
    ["stacked", /;\s*(drop|alter|create|insert|update|delete)\b/i],
    ["sleep", /\b(sleep|pg_sleep|benchmark)\s*\(/i]
];

function detectSqli(query) {
    const q = (query || "").trim();
    if (!q) return null;

    const hits = [];
    for (const [name, rx] of SQLI_REGEXES) {
        if (rx.test(q)) {
            hits.push(name);
        }
    }

    if (hits.length === 0) return null;

    return new Detection(
        "sql_injection",
        "SUSPICIOUS",
        7,
        "sql injection pattern detected",
        [...hits, "sql injection"]
    );
}

const PRIVATE_HOST_LITERALS = ["localhost", "127.0.0.1", "0.0.0.0"];
const METADATA_IPS = ["169.254.169.254"];

function isIpPrivate(ipStr) {
    const parts = ipStr.split('.');
    if (parts.length !== 4) return false;
    
    const p1 = parseInt(parts[0], 10);
    const p2 = parseInt(parts[1], 10);
    
    // 10.x.x.x
    if (p1 === 10) return true;
    // 172.16.x.x - 172.31.x.x
    if (p1 === 172 && p2 >= 16 && p2 <= 31) return true;
    // 192.168.x.x
    if (p1 === 192 && p2 === 168) return true;
    // Loopback 127.x.x.x
    if (p1 === 127) return true;
    // Link local 169.254.x.x
    if (p1 === 169 && p2 === 254) return true;
    
    return false;
}

function detectSsrf(url) {
    const u = (url || "").trim();
    if (!u) return null;

    const uLc = u.toLowerCase();

    if (uLc.startsWith("file://") || uLc.startsWith("gopher://") || uLc.startsWith("ftp://")) {
        return new Detection("ssrf", "SUSPICIOUS", 8, "dangerous outbound scheme (possible SSRF)", ["ssrf", "outbound request"]);
    }

    if (PRIVATE_HOST_LITERALS.some(h => uLc.includes(h))) {
        return new Detection("ssrf", "SUSPICIOUS", 8, "outbound request to localhost (possible SSRF)", ["ssrf", "localhost"]);
    }

    for (const ip of METADATA_IPS) {
        if (uLc.includes(ip)) {
            return new Detection("ssrf", "SUSPICIOUS", 9, "outbound request to instance metadata (possible SSRF)", ["ssrf", "metadata", ip]);
        }
    }

    const ipMatch = u.match(/\b(\d{1,3}(?:\.\d{1,3}){3})\b/);
    if (ipMatch && isIpPrivate(ipMatch[1])) {
        return new Detection("ssrf", "SUSPICIOUS", 8, "outbound request to private IP (possible SSRF)", ["ssrf", "private ip", ipMatch[1]]);
    }

    return null;
}

function clampSeverity(sev) {
    let v = parseInt(sev, 10);
    if (isNaN(v)) v = 1;
    return Math.max(1, Math.min(10, v));
}

module.exports = { Detection, detectSqli, detectSsrf, clampSeverity };
